core/src/main/scala/schema/provider/TroySchemaProvider.scala

```diff
@@ -19,35 +19,35 @@ package schema.provider
 
 import java.io.InputStream
 
-import cats.syntax.either._
+import cats.MonadError
 import freestyle.cassandra.schema._
 import troy.cql.ast.CqlParser
 
-class TroySchemaProvider(cqlF: => SchemaResult[String]) extends SchemaDefinitionProvider {
+class TroySchemaProvider[M[_]](cqlF: => M[String]) extends SchemaDefinitionProvider[M] {
 
-  override def schemaDefinition: SchemaResult[SchemaDefinition] =
-    cqlF.flatMap { cql =>
+  override def schemaDefinition(implicit M: MonadError[M, Throwable]): M[SchemaDefinition] =
+    M.flatMap(cqlF) { cql =>
       CqlParser.parseSchema(cql) match {
-        case CqlParser.Success(res, _) => Right(res)
+        case CqlParser.Success(res, _) => M.pure(res)
         case CqlParser.Failure(msg, next) =>
-          Left(
+          M.raiseError(
             SchemaDefinitionProviderError(
               s"Parse Failure: $msg, line = ${next.pos.line}, column = ${next.pos.column}"))
-        case CqlParser.Error(msg, _) => Left(SchemaDefinitionProviderError(msg))
+        case CqlParser.Error(msg, _) => M.raiseError(SchemaDefinitionProviderError(msg))
       }
     }
 }
 
 object TroySchemaProvider {
 
-  def apply(cql: String): TroySchemaProvider = new TroySchemaProvider(Right(cql))
+  def apply[M[_]](cql: String)(implicit M: MonadError[M, Throwable]): TroySchemaProvider[M] =
+    new TroySchemaProvider(M.pure(cql))
 
-  def apply(is: InputStream): TroySchemaProvider = new TroySchemaProvider(
-    Either.catchNonFatal {
-      scala.io.Source.fromInputStream(is).mkString
-    } leftMap { e =>
-      SchemaDefinitionProviderError(e.getMessage, Some(e))
-    }
-  )
+  def apply[M[_]](is: InputStream)(implicit M: MonadError[M, Throwable]): TroySchemaProvider[M] =
+    new TroySchemaProvider[M](
+      M.catchNonFatal {
+        scala.io.Source.fromInputStream(is).mkString
+      }
```

Comment:
You could still wrap the `Throwable` in a `SchemaDefinitionProviderError` if you wanted:

```scala
M.handleErrorWith(
  M.catchNonFatal(scala.io.Source.fromInputStream(is).mkString)
)(e => M.raiseError(SchemaDefinitionErrorProviderError(e.getMessage, Some(e)))
```